dashboard-widgets.el

```diff
@@ -302,84 +322,54 @@ WIDGET-PARAMS are passed to the \"widget-create\" function."
-                                  (eq (file-remote-p path) nil)
-                                  (file-directory-p path))
-                                 (cond
-                                  ((and (fboundp 'tramp-tramp-file-p)
-                                        (tramp-tramp-file-p default-directory))
-                                   (all-the-icons-octicon "file-directory"
-                                                          :height 1.0 :v-adjust 0.01))
-                                  ((file-symlink-p path)
-                                   (all-the-icons-octicon "file-symlink-directory"
-                                                          :height 1.0 :v-adjust 0.01))
-                                  ((all-the-icons-dir-is-submodule path)
-                                   (all-the-icons-octicon "file-submodule"
-                                                          :height 1.0 :v-adjust 0.01))
-                                  ((file-exists-p (format "%s/.git" path))
-                                   (all-the-icons-octicon "repo"
-                                                          :height 1.1 :v-adjust 0.01))
-                                  (t (let
-                                         ((matcher (all-the-icons-match-to-alist
-                                                    path all-the-icons-dir-icon-alist)))
-                                       (apply (car matcher) (list (cadr matcher)
-                                                                  :v-adjust 0.01)))))
-                               (cond
-                                ((string-equal ,section-name "Agenda for today:")
-                                 (all-the-icons-octicon "primitive-dot"
-                                                        :height 1.0 :v-adjust 0.01))
-                                ((eq (file-remote-p path) nil)
-                                 (all-the-icons-icon-for-file (file-name-nondirectory path)))
-                                (t (all-the-icons-octicon "radio-tower"
-                                                          :height 1.1 :v-adjust 0.01)))))))
-                 (insert "\t"))
-               (setq widget
-                     (widget-create 'push-button
-                                    :action ,action
-                                    :mouse-face 'highlight
-                                    :button-prefix ""
-                                    :button-suffix ""
-                                    :format "%[%t%]"
-                                    ,@rest))))
-           ,list)))
+     (mapc
+      (lambda (el)
+        (let ((tag ,@rest))
+          (insert "\n    ")
+
+          (when (and (display-graphic-p)
+                     dashboard-set-file-icons
+                     (featurep 'all-the-icons))
+            (let* ((path (car (last (split-string ,@rest " - "))))
+                   (icon (if (file-directory-p path)
```

Comment:
In line 340 `(if (and
                                  (eq (file-remote-p path) nil)
                                  (file-directory-p path))`
should be included instead of just (if (file-directory-p path) because for some reason the file-directory-p function is what establishes unnecesary tramp connections and sometimes deadlocks or crashes the dashboard. I realized this beacause  work with remote servers a lot and I discovered this was the solution.

Reply:
Good catch! I am not using remote server to test so didn't encounter this issue.

Reply:
Mr. seagle, I think it has to be `(eq (file-remote-p path) nil)` and not `(not (file-remote-p path)` because file-remote-p is not a boolean returning function! It returns a string if the remote server is available.

Reply:
Also in line 359 the `(file-remote-p path)` doesn't return true, it returns a string so the condition is never met

Reply:
@ema2159 I think  the check `(file-remote-p path)` is fine. The suffix `-p` implies return TRUE or FALSE. In docstring, it says `Test whether FILE specifies a location on a remote system`. Anyway, all strings including empty ("") is non-nil. So I don't see any risk here. Thoughts?

Reply:
It seems you're right.  I was confused because in the last cond statement you switched the two last conditions but I tried your code and it works!

Reply:
Great! Another thing is I got a better way to resolve the issue about remote file: `(file-directory-p (file-local-name path))` has a good performance. 
I didn't change since it looks good so far. You can test on your env.

Reply:
> Great! Another thing is I got a better way to resolve the issue about remote file: `(file-directory-p (file-local-name path))` has a good performance.
> I didn't change since it looks good so far. You can test on your env.

I think I already tried that and the problem is that when the remote isn't available it deadlocks (for example, I only have access to certain remotes in my office because of proxies and stuff, and when I try to get to them in my home it deadloks)

Reply:
Really? `file-local-name` won't trigger the remote connections IMO.

Reply:
Would you show me so I can try it out?

Reply:
Try this:

```elisp
(defmacro dashboard-insert-section-list (section-name list action &rest rest)
  "Insert into SECTION-NAME a LIST of items, expanding ACTION and passing REST to widget creation."
  `(when (car ,list)
     (mapc
      (lambda (el)
        (let ((tag ,@rest))
          (insert "\n    ")

          (when (and (display-graphic-p)
                     dashboard-set-file-icons
                     (featurep 'all-the-icons))
            (let* ((path (file-local-name (car (last (split-string ,@rest " - ")))))
                   (icon (if (file-directory-p path)
                             (cond
                              ((and (fboundp 'tramp-tramp-file-p)
                                    (tramp-tramp-file-p default-directory))
                               (all-the-icons-octicon "file-directory" :height 1.0 :v-adjust 0.01))
                              ((file-symlink-p path)
                               (all-the-icons-octicon "file-symlink-directory"
                                                      :height 1.0 :v-adjust 0.01))
                              ((all-the-icons-dir-is-submodule path)
                               (all-the-icons-octicon "file-submodule" :height 1.0 :v-adjust 0.01))
                              ((file-exists-p (format "%s/.git" path))
                               (all-the-icons-octicon "repo" :height 1.1 :v-adjust 0.01))
                              (t (let ((matcher (all-the-icons-match-to-alist
                                                 path all-the-icons-dir-icon-alist)))
                                   (apply (car matcher) (list (cadr matcher) :v-adjust 0.01)))))
                           (cond
                            ((string-equal ,section-name "Agenda for today:")
                             (all-the-icons-octicon "primitive-dot" :height 1.0 :v-adjust 0.01))
                            ((file-remote-p path)
                             (all-the-icons-octicon "radio-tower" :height 1.1 :v-adjust 0.01))
                            (t (all-the-icons-icon-for-file (file-name-nondirectory path)))))))
              (setq tag (concat icon " " ,@rest))))

          (widget-create 'item
                         :tag tag
                         :action ,action
                         :mouse-face 'highlight
                         :button-prefix ""
                         :button-suffix ""
                         :format "%[%t%]")))
      ,list)))
```

Reply:
Ok! Let me give it a try 😁

Reply:
@seagle0128 No Mr. Zhang, for some reason this happens (this is what happened to me before also):
![image](https://user-images.githubusercontent.com/17773218/58426185-05d80a80-8059-11e9-99f0-5c5bd974d74e.png)
In my bookmarks there is a remote path and for some reason that implementation makes that error happen! I think it's better to leave it as it is right now. Although there are some other Centaur Emacs features that could be very nice to have like the footnote or the buttons. Although I'm not really sure what the latter would be useful for, it coul be very nice that the user was able to insert buttons with links and stuff and I think they look awesome!

Reply:
@ema2159 I guess you are using Emacs  25.1 or 25.2. `file-local-name` is introduced in 25.2.2. It also leverage `file-remote-p` to avoid unnecessary remote connections, but makes the codes more clear.  Of course, it's fine to leave it as it is now.

```elisp
;;`file-local-name' is introduced in 25.2.2.
(unless (fboundp 'file-local-name)
  (defun file-local-name (file)
    "Return the local name component of FILE.
It returns a file name which can be used directly as argument of
`process-file', `start-file-process', or `shell-command'."
    (or (file-remote-p file 'localname) file)))
```

Reply:
No sir! I'm using emacs 26. It's really weird tho, because your implementation shouldn't give any problems.

Reply:
You can look into the back traces and do some troubleshooting. It's easy to fix I think.